packages/actor-optimize-query-operation-group-sources/lib/ActorOptimizeQueryOperationGroupSources.ts

```diff
@@ -179,11 +175,23 @@ export class ActorOptimizeQueryOperationGroupSources extends ActorOptimizeQueryO
     source: IQuerySourceWrapper | undefined,
     context: IActionContext,
   ): Promise<O> {
-    if (source && doesShapeAcceptOperation(await source.source.getSelectorShape(context), operation)) {
-      this.logDebug(context, `Hoist ${inputs.length} source-specific operations into a single ${operation.type} operation for ${source.source.toString()}`);
-      operation = assignOperationSource(operation, source);
-      for (const input of inputs) {
-        removeOperationSource(input);
+    if (source) {
+      const shape: FragmentSelectorShape = await source.source.getSelectorShape(context);
+      if (doesShapeAcceptOperation(shape, operation)) {
+        // Checks for extension functions:
+        // The first 3 operands in the OR are checks to see if it's an extension function comunica supports
+        // If not, then we're good to go
+        // If it does, then only move the source annotation upwards if the source also supports the extension function
+        const extensionFunctions = context.get(KeysInitQuery.extensionFunctions);
+        const expression: Algebra.Operation | undefined = operation.expression;
```

Comment:
We would not need the cast if we would first check the operation type, to ensure it's an expression.